{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "#Initial Imports\n",
    "import os#; print('os version: ', os.__version__)\n",
    "import shutil#; print('shutil version: ', shutil.__version__)\n",
    "from subprocess import PIPE, Popen\n",
    "from fake_useragent import UserAgent\n",
    "import os.path\n",
    "from os import path\n",
    "from Bio.ExPASy import Enzyme\n",
    "import numpy as np \n",
    "from sklearn.metrics import pairwise_distances\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "red = \"\\033[91m\"\n",
    "reset_color = \"\\033[0m\"\n",
    "import tkinter as tk\n",
    "from tkinter import filedialog\n",
    "from tkinter import ttk\n",
    "\n",
    "from IW_functions import EC_extract, tsv_to_fasta, diamond_impl, genome_extractor, genome_to_genome_diffcomp, read_in_binary_matrix, calculating_distance, pass_to_distance\n",
    "\n",
    "#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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to Environmental Census (EnCen), a bioinformatics tool for characterizing synthetic biological risk to the environment \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('Welcome to Environmental Census (EnCen), a bioinformatics tool for characterizing synthetic biological risk to the environment \\n')\n",
    "# EC_extract()\n",
    "# tsv_to_fasta()\n",
    "mg_to_analyze = input('Please input the metagenome you would like to analyze:')\n",
    "#options are industrial wastewater, WWTP, and River "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Industrial Wastewater\n"
     ]
    }
   ],
   "source": [
    "print(mg_to_analyze) \n",
    "\n",
    "#Will probably have to add exception handling into this \n",
    "#Add file to analyze upload here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Full Calling Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anna/Documents/JGI_soil_genomes/Industrial Wastewater_metagenome_bins\n",
      "DIAMOND library is located in:  /home/anna/Documents/JGI_soil_genomes/Industrial Wastewater_metagenome_bins\n",
      "Creation of DIAMOND-formatted library...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "diamond v0.9.30.131 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "\n",
      "#CPU threads: 20\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Database file: /home/anna/Documents/JGI_soil_genomes/uniprot.fasta\n",
      "Opening the database file...  [0s]\n",
      "Loading sequences...  [0.4s]\n",
      "Masking sequences...  [0.625s]\n",
      "Writing sequences...  [0.1s]\n",
      "Hashing sequences...  [0.038s]\n",
      "Loading sequences...  [0s]\n",
      "Writing trailer...  [0.003s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the database file...  [0s]\n",
      "Database hash = 9ae9565aeffc045297d8444896762120\n",
      "Processed 275960 sequences, 114276793 letters.\n",
      "Total time = 1.168s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Library complete\n",
      "['3300000513_1 (8th copy)', 'faa']\n",
      "Processing  3300000513_1 (8th copy)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "diamond v0.9.30.131 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "\n",
      "#CPU threads: 20\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: \n",
      "Opening the database...  [0s]\n",
      "#Target sequences to report alignments for: 1\n",
      "Reference = Uniprot_Reference_Library.dmnd\n",
      "Sequences = 275960\n",
      "Letters = 114276793\n",
      "Block size = 2000000000\n",
      "Opening the input file...  [0s]\n",
      "Opening the output file...  [0s]\n",
      "Loading query sequences...  [0.001s]\n",
      "Masking queries...  [0.007s]\n",
      "Building query seed set...  [0.003s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0.001s]\n",
      "Allocating buffers...  [0s]\n",
      "Loading reference sequences...  [0.15s]\n",
      "Masking reference...  [0.597s]\n",
      "Initializing temporary storage...  [0s]\n",
      "Building reference histograms...  [0.252s]\n",
      "Allocating buffers...  [0s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 0.\n",
      "Building reference seed array...  [0.172s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.026s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.014s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 1.\n",
      "Building reference seed array...  [0.193s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.026s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.01s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 2.\n",
      "Building reference seed array...  [0.185s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.027s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.015s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 3.\n",
      "Building reference seed array...  [0.153s]\n",
      "Building query seed array...  [0.002s]\n",
      "Computing hash join...  [0.028s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.011s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 0.\n",
      "Building reference seed array...  [0.15s]\n",
      "Building query seed array...  [0.002s]\n",
      "Computing hash join...  [0.028s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.01s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 1.\n",
      "Building reference seed array...  [0.194s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.028s]\n",
      "Building seed filter...  [0s]\n",
      "Searching alignments...  [0.009s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 2.\n",
      "Building reference seed array...  [0.191s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.028s]\n",
      "Building seed filter...  [0s]\n",
      "Searching alignments...  [0.011s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 3.\n",
      "Building reference seed array...  [0.151s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.028s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.011s]\n",
      "Deallocating buffers...  [0.019s]\n",
      "Computing alignments...  [0.283s]\n",
      "Deallocating reference...  [0.007s]\n",
      "Loading reference sequences...  [0s]\n",
      "Deallocating buffers...  [0s]\n",
      "Deallocating queries...  [0s]\n",
      "Loading query sequences...  [0s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the output file...  [0s]\n",
      "Closing the database file...  [0s]\n",
      "Deallocating taxonomy...  [0s]\n",
      "Total time = 3.063s\n",
      "Reported 672 pairwise alignments, 672 HSPs.\n",
      "672 queries aligned.\n",
      "diamond v0.9.30.131 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "\n",
      "#CPU threads: 20\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Database file: /home/anna/Documents/JGI_soil_genomes/uniprot.fasta\n",
      "Opening the database file...  [0.015s]\n",
      "Loading sequences... "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diamond_impl--success\n",
      "/home/anna/Documents/JGI_soil_genomes/reference_diamond_analysis_output\n",
      "3300000513_1 (8th copy)_matches.tsv\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "File already Exists\n",
      "/home/anna/Documents/JGI_soil_genomes/reference_diamond_analysis_output/Industrial Wastewater_metagenome_functional_profile\n",
      "Synbio directory already exists\n",
      "\u001b[91mATTENTION USER: SYNBIO OR COMPARISON ORGANISM .FAA MUST BE IN THE SYNBIO INPUTS AND OUTPUTS FOLDER. \n",
      " OTHERWISE, SCRIPT WILL CRASH\u001b[0m\n",
      "/home/anna/Documents/JGI_soil_genomes/Industrial Wastewater_synbio_inputs_and_outputs\n",
      "DIAMOND library is located in:  /home/anna/Documents/JGI_soil_genomes/Industrial Wastewater_synbio_inputs_and_outputs\n",
      "Creation of DIAMOND-formatted library...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " [0.383s]\n",
      "Masking sequences...  [0.727s]\n",
      "Writing sequences...  [0.105s]\n",
      "Hashing sequences...  [0.043s]\n",
      "Loading sequences...  [0s]\n",
      "Writing trailer...  [0.003s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the database file...  [0.049s]\n",
      "Database hash = 9ae9565aeffc045297d8444896762120\n",
      "Processed 275960 sequences, 114276793 letters.\n",
      "Total time = 1.327s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Library complete\n",
      "Processing  Synbio\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "diamond v0.9.30.131 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "\n",
      "#CPU threads: 20\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: \n",
      "Opening the database...  [0s]\n",
      "#Target sequences to report alignments for: 1\n",
      "Reference = Uniprot_Reference_Library.dmnd\n",
      "Sequences = 275960\n",
      "Letters = 114276793\n",
      "Block size = 2000000000\n",
      "Opening the input file...  [0s]\n",
      "Opening the output file...  [0s]\n",
      "Loading query sequences...  [0.001s]\n",
      "Masking queries...  [0.007s]\n",
      "Building query seed set...  [0.003s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0.001s]\n",
      "Allocating buffers...  [0s]\n",
      "Loading reference sequences...  [0.138s]\n",
      "Masking reference...  [0.887s]\n",
      "Initializing temporary storage...  [0s]\n",
      "Building reference histograms...  [0.294s]\n",
      "Allocating buffers...  [0s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 0.\n",
      "Building reference seed array...  [0.204s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.033s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.012s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 1.\n",
      "Building reference seed array...  [0.242s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.032s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.009s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 2.\n",
      "Building reference seed array...  [0.245s]\n",
      "Building query seed array...  [0.002s]\n",
      "Computing hash join...  [0.03s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.01s]\n",
      "Processing query block 0, reference block 0, shape 0, index chunk 3.\n",
      "Building reference seed array...  [0.197s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.032s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.009s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 0.\n",
      "Building reference seed array...  [0.192s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.03s]\n",
      "Building seed filter...  [0s]\n",
      "Searching alignments...  [0.008s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 1.\n",
      "Building reference seed array...  [0.242s]\n",
      "Building query seed array...  [0.001s]\n",
      "Computing hash join...  [0.032s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.008s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 2.\n",
      "Building reference seed array...  [0.233s]\n",
      "Building query seed array...  [0.002s]\n",
      "Computing hash join...  [0.034s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.007s]\n",
      "Processing query block 0, reference block 0, shape 1, index chunk 3.\n",
      "Building reference seed array...  [0.196s]\n",
      "Building query seed array...  [0.002s]\n",
      "Computing hash join...  [0.034s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.008s]\n",
      "Deallocating buffers...  [0.023s]\n",
      "Computing alignments...  [0.255s]\n",
      "Deallocating reference...  [0.007s]\n",
      "Loading reference sequences...  [0s]\n",
      "Deallocating buffers...  [0s]\n",
      "Deallocating queries...  [0s]\n",
      "Loading query sequences...  [0s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the output file...  [0s]\n",
      "Closing the database file...  [0s]\n",
      "Deallocating taxonomy...  [0s]\n",
      "Total time = 3.744s\n",
      "Reported 775 pairwise alignments, 775 HSPs.\n",
      "775 queries aligned.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diamond_impl--success\n",
      "/home/anna/Documents/JGI_soil_genomes/Industrial Wastewater_synbio_inputs_and_outputs\n",
      "Synbio_matches.tsv\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "Appending...\n",
      "File already Exists\n",
      "/home/anna/Documents/JGI_soil_genomes/Industrial Wastewater_synbio_inputs_and_outputs/Synbio_functional_profile\n",
      "Synbio  size of  (1, 8236)  successfully imported.\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/home/anna/Documents/JGI_soil_genomes/functional_profiles/Industrial Wastewater/_metagenome_functional_profile'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 97\u001b[0m\n\u001b[1;32m     95\u001b[0m \u001b[38;5;66;03m#____________________________________________________________________________________#Distance Scoring\u001b[39;00m\n\u001b[1;32m     96\u001b[0m     synbio_binary \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/home/anna/Documents/JGI_soil_genomes/functional_profiles/Synbio_functional_profile\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m---> 97\u001b[0m     [distance_list_for_synbio, new_loc ]\u001b[38;5;241m=\u001b[39m pass_to_distance(synbio_binary, name, desired_location2, mg_to_analyze)\n\u001b[1;32m     98\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSynbio Analysis Complete\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    100\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m mg_to_analyze \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mWWTP\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "File \u001b[0;32m~/Desktop/JD_Niche_OverLap (Git)/Niche_JD/Eco_V2/EnCen/MultiMetagenome/IW_functions.py:262\u001b[0m, in \u001b[0;36mpass_to_distance\u001b[0;34m(synbio_binary, sb_name, desired_location, mg_to_analyze)\u001b[0m\n\u001b[1;32m    261\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpass_to_distance\u001b[39m(synbio_binary, sb_name, desired_location, mg_to_analyze):\n\u001b[0;32m--> 262\u001b[0m     all_matrix \u001b[38;5;241m=\u001b[39m read_in_binary_matrix(synbio_binary, sb_name, mg_to_analyze)\n\u001b[1;32m    263\u001b[0m     list_genomes \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(all_matrix\u001b[38;5;241m.\u001b[39mindex)\n\u001b[1;32m    264\u001b[0m     \u001b[38;5;66;03m# Completes distance calculation on an Euclidean basis. Returns the synbio column with genome names appended\u001b[39;00m\n",
      "File \u001b[0;32m~/Desktop/JD_Niche_OverLap (Git)/Niche_JD/Eco_V2/EnCen/MultiMetagenome/IW_functions.py:202\u001b[0m, in \u001b[0;36mread_in_binary_matrix\u001b[0;34m(synbio_binary, sb_name, mg_to_analyze)\u001b[0m\n\u001b[1;32m    200\u001b[0m \u001b[38;5;28mprint\u001b[39m(sb_name, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m size of \u001b[39m\u001b[38;5;124m\"\u001b[39m, np\u001b[38;5;241m.\u001b[39mshape(synbio_binary), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m successfully imported.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    201\u001b[0m \u001b[38;5;66;03m# Opens the matrix that includes the Bacteria and Archaea summary result\u001b[39;00m\n\u001b[0;32m--> 202\u001b[0m domain_binary \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/home/anna/Documents/JGI_soil_genomes/functional_profiles/\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m mg_to_analyze \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_metagenome_functional_profile\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m    203\u001b[0m                               delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m\"\u001b[39m, header\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m    204\u001b[0m \u001b[38;5;66;03m#this is from chunk 1, and is the EC_Binary we generated earlier\u001b[39;00m\n\u001b[1;32m    205\u001b[0m domain_binary \u001b[38;5;241m=\u001b[39m domain_binary\u001b[38;5;241m.\u001b[39mset_index(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mName_of_MetaGenome_Bin\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_engine(f, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m get_handle(\n\u001b[1;32m   1662\u001b[0m     f,\n\u001b[1;32m   1663\u001b[0m     mode,\n\u001b[1;32m   1664\u001b[0m     encoding\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1665\u001b[0m     compression\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcompression\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1666\u001b[0m     memory_map\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmemory_map\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m),\n\u001b[1;32m   1667\u001b[0m     is_text\u001b[38;5;241m=\u001b[39mis_text,\n\u001b[1;32m   1668\u001b[0m     errors\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding_errors\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstrict\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[1;32m   1669\u001b[0m     storage_options\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstorage_options\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1670\u001b[0m )\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(\n\u001b[1;32m    860\u001b[0m             handle,\n\u001b[1;32m    861\u001b[0m             ioargs\u001b[38;5;241m.\u001b[39mmode,\n\u001b[1;32m    862\u001b[0m             encoding\u001b[38;5;241m=\u001b[39mioargs\u001b[38;5;241m.\u001b[39mencoding,\n\u001b[1;32m    863\u001b[0m             errors\u001b[38;5;241m=\u001b[39merrors,\n\u001b[1;32m    864\u001b[0m             newline\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    865\u001b[0m         )\n\u001b[1;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/home/anna/Documents/JGI_soil_genomes/functional_profiles/Industrial Wastewater/_metagenome_functional_profile'"
     ]
    }
   ],
   "source": [
    "##_______________________________________________________________________________## initial def's \n",
    "if mg_to_analyze == 'Industrial Wastewater':\n",
    "    metagenome_name = 'reference_diamond_analysis_output' #-> folder\n",
    "    home_dir = '/home/anna/Documents/JGI_soil_genomes' \n",
    "    # IW = '/home/anna/Documents/JGI_soil_genomes/IW_Metagenome'\n",
    "    IW = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_metagenome_bins'\n",
    "    # abspath = os.path.abspath()\n",
    "#Automatically Makes a folder for the metagenomic bins to upload into, then asks the user for files and moves those files into the bins folder_____________________________________________________###\n",
    "    os.chdir(home_dir)\n",
    "    # if os.path.exists(IW):\n",
    "    #     shutil.rmtree(IW)\n",
    "    #     os.mkdir(IW)\n",
    "    # else:\n",
    "    #     os.mkdir(IW)\n",
    "\n",
    "    def upload_file():\n",
    "        root = tk.Tk()\n",
    "        root.withdraw() # Hide the main window\n",
    "        file_paths = filedialog.askopenfilenames(initialdir= home_dir, title = 'Select .faa files you would like to compare to')\n",
    "        return file_paths\n",
    "    \n",
    "    def move_files_to_folder(file_paths, destination_folder):\n",
    "        for file_path in file_paths:\n",
    "            if os.path.isfile(file_path):\n",
    "                shutil.move(file_path, destination_folder)\n",
    "\n",
    "    # # Example usage\n",
    "    file_paths = upload_file()\n",
    "    upload_location = IW\n",
    "    move_files_to_folder(file_paths, upload_location)\n",
    "\n",
    "##__________________________________________________________________________________#Diamond Analysis\n",
    "    os.chdir(home_dir)\n",
    "    if os.path.exists(metagenome_name):\n",
    "        shutil.rmtree(metagenome_name)\n",
    "        os.mkdir(metagenome_name)\n",
    "    else:\n",
    "        os.mkdir(metagenome_name)\n",
    "    # home_dir = home_dir + \"/\" + metagenome_name \n",
    "\n",
    "    os.chdir(IW)\n",
    "    diamond = diamond_impl(IW, '') #-> Takes in the path and directory\n",
    "# #________________________________________________________________________________# Creating reference functional profile\n",
    "\n",
    "    dmnd_folder = '/home/anna/Documents/JGI_soil_genomes/reference_diamond_analysis_output'\n",
    "    ff_name = 'functional_profiles'\n",
    "    functional_folder = '/home/anna/Documents/JGI_soil_genomes/functional_profiles'\n",
    "    name = mg_to_analyze + '_metagenome'\n",
    "\n",
    "    for item in os.listdir(IW):\n",
    "        if item.endswith(('_matches.tsv', '.dmnd')):\n",
    "            source = os.path.join(IW, item)\n",
    "            destination = os.path.join(dmnd_folder, item)\n",
    "            shutil.move(source, destination)\n",
    "    \n",
    "\n",
    "    output = genome_extractor(dmnd_folder, name, home_dir)\n",
    "\n",
    "    os.chdir(home_dir)\n",
    "    if os.path.exists(functional_folder):\n",
    "        shutil.rmtree(functional_folder)\n",
    "        os.mkdir(ff_name)\n",
    "    else:#makes a new directory called metagenome_name\n",
    "        os.mkdir(ff_name)\n",
    "\n",
    "    for item in os.listdir(dmnd_folder):\n",
    "        if item.endswith('_profile'):\n",
    "            source = os.path.join(dmnd_folder, item)\n",
    "            destination = os.path.join(functional_folder, item)\n",
    "            shutil.move(source, destination)\n",
    "        \n",
    "#-----------------------------------------------------------------------------------# Organism to analyze matches and functional profile creation \n",
    "    synbio = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "    name = 'Synbio'\n",
    "    syn_folder_name = mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "    desired_location2 = '/home/anna/Documents/JGI_soil_genomes'\n",
    "\n",
    "    os.chdir(home_dir)\n",
    "    if os.path.exists(synbio):\n",
    "        print('Synbio directory already exists')\n",
    "    else:#makes a new directory called metagenome_name\n",
    "        os.mkdir(syn_folder_name)\n",
    "\n",
    "    print(f'{red}ATTENTION USER: SYNBIO OR COMPARISON ORGANISM .FAA MUST BE IN THE SYNBIO INPUTS AND OUTPUTS FOLDER. \\n OTHERWISE, SCRIPT WILL CRASH{reset_color}')\n",
    "    os.chdir(synbio) \n",
    "    diamond_syn = diamond_impl(synbio, name) #diamond_syn = synbio\n",
    "    output2 = genome_extractor(diamond_syn, name, home_dir)\n",
    "\n",
    "    for item in os.listdir(synbio):\n",
    "        if item.endswith('_profile'):\n",
    "            source = os.path.join(synbio, item)\n",
    "            destination = os.path.join(functional_folder, item)\n",
    "            shutil.move(source, destination)\n",
    "\n",
    "#____________________________________________________________________________________#Distance Scoring\n",
    "    synbio_binary = '/home/anna/Documents/JGI_soil_genomes/functional_profiles/Synbio_functional_profile'\n",
    "    [distance_list_for_synbio, new_loc ]= pass_to_distance(synbio_binary, name, desired_location2, mg_to_analyze)\n",
    "    print('Synbio Analysis Complete')\n",
    "\n",
    "elif mg_to_analyze == 'WWTP':\n",
    "    print('works')\n",
    "else: \n",
    "    print('\"They say that there\\'s no medicine that can cure a fool. I guess that\\'s true ...\" - Potion Shop Hag')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
