{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To Do \n",
    "- "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "#Initial Imports\n",
    "import os#; print('os version: ', os.__version__)\n",
    "import shutil#; print('shutil version: ', shutil.__version__)\n",
    "from subprocess import PIPE, Popen\n",
    "from fake_useragent import UserAgent\n",
    "import os.path\n",
    "from os import path\n",
    "from Bio.ExPASy import Enzyme\n",
    "import numpy as np \n",
    "from sklearn.metrics import pairwise_distances\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "red = \"\\033[91m\"\n",
    "reset_color = \"\\033[0m\"\n",
    "import tkinter as tk\n",
    "from tkinter import filedialog\n",
    "from tkinter import ttk\n",
    "\n",
    "from IW_functions import EC_extract, tsv_to_fasta, diamond_impl, genome_extractor, genome_to_genome_diffcomp, read_in_binary_matrix, calculating_distance, pass_to_distance, upload_file, upload_file2, move_files_to_folder\n",
    "\n",
    "#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a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to Environmental Census (EnCen), a bioinformatics tool for characterizing synthetic biological risk to the environment \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('Welcome to Environmental Census (EnCen), a bioinformatics tool for characterizing synthetic biological risk to the environment \\n')\n",
    "# EC_extract()\n",
    "# tsv_to_fasta()\n",
    "intake = input('Please input the metagenomes you would like to analyze(Industrial Wastewater, WWTP, or River), seperated by commas: ')\n",
    "choices = [choice.strip() for choice in intake.split(\",\")]\n",
    "\n",
    "#options are industrial wastewater, WWTP, and River "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['wwtp']\n"
     ]
    }
   ],
   "source": [
    "print(choices) \n",
    "#Will probably have to add exception handling into this \n",
    "#Add file to analyze upload here so it can be referenced in the subsequent part. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Full Calling Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mg_to_analyze in choices:\n",
    "    if mg_to_analyze == 'industrial wastewater':\n",
    "        metagenome_name = 'reference_diamond_analysis_output' #-> folder\n",
    "        home_dir = '/home/anna/Documents/JGI_soil_genomes' \n",
    "        # IW = '/home/anna/Documents/JGI_soil_genomes/IW_Metagenome'\n",
    "        IW = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_metagenome_bins'\n",
    "        # abspath = os.path.abspath()\n",
    "    #Automatically Makes a folder for the metagenomic bins to upload into, then asks the user for files and moves those files into the bins folder_____________________________________________________###\n",
    "        # os.chdir(home_dir)\n",
    "        # if os.path.exists(IW):\n",
    "        #     shutil.rmtree(IW)\n",
    "        #     os.mkdir(IW)\n",
    "        # else:\n",
    "        #     os.mkdir(IW)\n",
    "\n",
    "        # # Example usage\n",
    "        file_paths = upload_file(home_dir, mg_to_analyze)\n",
    "        upload_location = IW\n",
    "        move_files_to_folder(file_paths, upload_location)\n",
    "\n",
    "    ##__________________________________________________________________________________#Diamond Analysis\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(metagenome_name):\n",
    "            shutil.rmtree(metagenome_name)\n",
    "            os.mkdir(metagenome_name)\n",
    "        else:\n",
    "            os.mkdir(metagenome_name)\n",
    "        # home_dir = home_dir + \"/\" + metagenome_name \n",
    "\n",
    "        os.chdir(IW)\n",
    "        diamond = diamond_impl(IW, '') #-> Takes in the path and directory\n",
    "    # #________________________________________________________________________________# Creating reference functional profile\n",
    "\n",
    "        dmnd_folder = '/home/anna/Documents/JGI_soil_genomes/reference_diamond_analysis_output'\n",
    "        ff_name = 'functional_profiles'\n",
    "        functional_folder = '/home/anna/Documents/JGI_soil_genomes/functional_profiles'\n",
    "        name = mg_to_analyze + '_metagenome'\n",
    "\n",
    "        for item in os.listdir(IW):\n",
    "            if item.endswith(('_matches.tsv', '.dmnd')):\n",
    "                source = os.path.join(IW, item)\n",
    "                destination = os.path.join(dmnd_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "        \n",
    "\n",
    "        output = genome_extractor(dmnd_folder, name, home_dir)\n",
    "\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(functional_folder):\n",
    "            shutil.rmtree(functional_folder)\n",
    "            os.mkdir(ff_name)\n",
    "        else:#makes a new directory called metagenome_name\n",
    "            os.mkdir(ff_name)\n",
    "\n",
    "        for item in os.listdir(dmnd_folder):\n",
    "            if item.endswith('_profile'):\n",
    "                source = os.path.join(dmnd_folder, item)\n",
    "                destination = os.path.join(functional_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "            \n",
    "    #-----------------------------------------------------------------------------------# Organism to analyze matches and functional profile creation \n",
    "        synbio = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "        name = 'Synbio'\n",
    "        syn_folder_name = mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "        desired_location2 = '/home/anna/Documents/JGI_soil_genomes'\n",
    "\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(synbio):\n",
    "            print('Synbio directory already exists')\n",
    "        else:#makes a new directory called metagenome_name\n",
    "            os.mkdir(syn_folder_name)\n",
    "\n",
    "       \n",
    "        \n",
    "        file_paths = upload_file2(home_dir, mg_to_analyze)\n",
    "        upload_location = synbio\n",
    "        move_files_to_folder(file_paths, upload_location)\n",
    "\n",
    "        os.chdir(synbio) \n",
    "        diamond_syn = diamond_impl(synbio, name) #diamond_syn = synbio\n",
    "        output2 = genome_extractor(diamond_syn, name, home_dir)\n",
    "\n",
    "        for item in os.listdir(synbio):\n",
    "            if item.endswith('_profile'):\n",
    "                source = os.path.join(synbio, item)\n",
    "                destination = os.path.join(functional_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "\n",
    "    #____________________________________________________________________________________#Distance Scoring\n",
    "        synbio_binary = '/home/anna/Documents/JGI_soil_genomes/functional_profiles/Synbio_functional_profile'\n",
    "        [distance_list_for_synbio, new_loc ]= pass_to_distance(synbio_binary, name, desired_location2, mg_to_analyze)\n",
    "        print(mg_to_analyze + f'{red} vs. Synbio Analysis Complete{reset_color}')\n",
    "\n",
    "    elif mg_to_analyze == 'wwtp':\n",
    "        metagenome_name = 'reference_diamond_analysis_output' #-> folder\n",
    "        home_dir = '/home/anna/Documents/JGI_soil_genomes' \n",
    "        # IW = '/home/anna/Documents/JGI_soil_genomes/IW_Metagenome'\n",
    "        IW = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_metagenome_bins'\n",
    "        # abspath = os.path.abspath()\n",
    "    #Automatically Makes a folder for the metagenomic bins to upload into, then asks the user for files and moves those files into the bins folder_____________________________________________________###\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(IW):\n",
    "            shutil.rmtree(IW)\n",
    "            os.mkdir(IW)\n",
    "        else:\n",
    "            os.mkdir(IW)\n",
    "\n",
    "        # # Example usage\n",
    "        file_paths = upload_file(home_dir, mg_to_analyze)\n",
    "        upload_location = IW\n",
    "        move_files_to_folder(file_paths, upload_location)\n",
    "\n",
    "    ##__________________________________________________________________________________#Diamond Analysis\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(metagenome_name):\n",
    "            shutil.rmtree(metagenome_name)\n",
    "            os.mkdir(metagenome_name)\n",
    "        else:\n",
    "            os.mkdir(metagenome_name)\n",
    "        # home_dir = home_dir + \"/\" + metagenome_name \n",
    "\n",
    "        os.chdir(IW)\n",
    "        diamond = diamond_impl(IW, '') #-> Takes in the path and directory\n",
    "    # #________________________________________________________________________________# Creating reference functional profile\n",
    "\n",
    "        dmnd_folder = '/home/anna/Documents/JGI_soil_genomes/reference_diamond_analysis_output'\n",
    "        ff_name = 'functional_profiles'\n",
    "        functional_folder = '/home/anna/Documents/JGI_soil_genomes/functional_profiles'\n",
    "        name = mg_to_analyze + '_metagenome'\n",
    "\n",
    "        for item in os.listdir(IW):\n",
    "            if item.endswith(('_matches.tsv', '.dmnd')):\n",
    "                source = os.path.join(IW, item)\n",
    "                destination = os.path.join(dmnd_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "        \n",
    "\n",
    "        output = genome_extractor(dmnd_folder, name, home_dir)\n",
    "\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(functional_folder):\n",
    "            shutil.rmtree(functional_folder)\n",
    "            os.mkdir(ff_name)\n",
    "        else:#makes a new directory called metagenome_name\n",
    "            os.mkdir(ff_name)\n",
    "\n",
    "        for item in os.listdir(dmnd_folder):\n",
    "            if item.endswith('_profile'):\n",
    "                source = os.path.join(dmnd_folder, item)\n",
    "                destination = os.path.join(functional_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "            \n",
    "    #-----------------------------------------------------------------------------------# Organism to analyze matches and functional profile creation \n",
    "        synbio = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "        name = 'Synbio'\n",
    "        syn_folder_name = mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "        desired_location2 = '/home/anna/Documents/JGI_soil_genomes'\n",
    "\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(synbio):\n",
    "            print('Synbio directory already exists')\n",
    "        else:#makes a new directory called metagenome_name\n",
    "            os.mkdir(syn_folder_name)\n",
    "\n",
    "        file_paths = upload_file2(home_dir, mg_to_analyze)\n",
    "        upload_location = synbio\n",
    "        move_files_to_folder(file_paths, upload_location)\n",
    "\n",
    "        os.chdir(synbio) \n",
    "        diamond_syn = diamond_impl(synbio, name) #diamond_syn = synbio\n",
    "        output2 = genome_extractor(diamond_syn, name, home_dir)\n",
    "\n",
    "        for item in os.listdir(synbio):\n",
    "            if item.endswith('_profile'):\n",
    "                source = os.path.join(synbio, item)\n",
    "                destination = os.path.join(functional_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "                \n",
    "        synbio_binary = '/home/anna/Documents/JGI_soil_genomes/functional_profiles/Synbio_functional_profile'\n",
    "        [distance_list_for_synbio, new_loc ]= pass_to_distance(synbio_binary, name, desired_location2, mg_to_analyze)\n",
    "        print(mg_to_analyze + 'vs. Synbio Analysis Complete')\n",
    "\n",
    "    elif mg_to_analyze == 'river':\n",
    "        metagenome_name = 'reference_diamond_analysis_output' #-> folder\n",
    "        home_dir = '/home/anna/Documents/JGI_soil_genomes' \n",
    "        # IW = '/home/anna/Documents/JGI_soil_genomes/IW_Metagenome'\n",
    "        IW = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_metagenome_bins'\n",
    "        # abspath = os.path.abspath()\n",
    "    #Automatically Makes a folder for the metagenomic bins to upload into, then asks the user for files and moves those files into the bins folder_____________________________________________________###\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(IW):\n",
    "            shutil.rmtree(IW)\n",
    "            os.mkdir(IW)\n",
    "        else:\n",
    "            os.mkdir(IW)\n",
    "\n",
    "        # # Example usage\n",
    "        file_paths = upload_file(home_dir, mg_to_analyze)\n",
    "        upload_location = IW\n",
    "        move_files_to_folder(file_paths, upload_location)\n",
    "\n",
    "    ##__________________________________________________________________________________#Diamond Analysis\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(metagenome_name):\n",
    "            shutil.rmtree(metagenome_name)\n",
    "            os.mkdir(metagenome_name)\n",
    "        else:\n",
    "            os.mkdir(metagenome_name)\n",
    "        # home_dir = home_dir + \"/\" + metagenome_name \n",
    "\n",
    "        os.chdir(IW)\n",
    "        diamond = diamond_impl(IW, '') #-> Takes in the path and directory\n",
    "    # #________________________________________________________________________________# Creating reference functional profile\n",
    "\n",
    "        dmnd_folder = '/home/anna/Documents/JGI_soil_genomes/reference_diamond_analysis_output'\n",
    "        ff_name = 'functional_profiles'\n",
    "        functional_folder = '/home/anna/Documents/JGI_soil_genomes/functional_profiles'\n",
    "        name = mg_to_analyze + '_metagenome'\n",
    "\n",
    "        for item in os.listdir(IW):\n",
    "            if item.endswith(('_matches.tsv', '.dmnd')):\n",
    "                source = os.path.join(IW, item)\n",
    "                destination = os.path.join(dmnd_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "        \n",
    "\n",
    "        output = genome_extractor(dmnd_folder, name, home_dir)\n",
    "\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(functional_folder):\n",
    "            shutil.rmtree(functional_folder)\n",
    "            os.mkdir(ff_name)\n",
    "        else:#makes a new directory called metagenome_name\n",
    "            os.mkdir(ff_name)\n",
    "\n",
    "        for item in os.listdir(dmnd_folder):\n",
    "            if item.endswith('_profile'):\n",
    "                source = os.path.join(dmnd_folder, item)\n",
    "                destination = os.path.join(functional_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "            \n",
    "    #-----------------------------------------------------------------------------------# Organism to analyze matches and functional profile creation \n",
    "        synbio = '/home/anna/Documents/JGI_soil_genomes/' + mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "        name = 'Synbio'\n",
    "        syn_folder_name = mg_to_analyze + '_synbio_inputs_and_outputs'\n",
    "        desired_location2 = '/home/anna/Documents/JGI_soil_genomes'\n",
    "\n",
    "        os.chdir(home_dir)\n",
    "        if os.path.exists(synbio):\n",
    "            print('Synbio directory already exists')\n",
    "        else:#makes a new directory called metagenome_name\n",
    "            os.mkdir(syn_folder_name)\n",
    "\n",
    "        file_paths = upload_file2(home_dir, mg_to_analyze)\n",
    "        upload_location = synbio\n",
    "        move_files_to_folder(file_paths, upload_location)\n",
    "\n",
    "        os.chdir(synbio) \n",
    "        diamond_syn = diamond_impl(synbio, name) #diamond_syn = synbio\n",
    "        output2 = genome_extractor(diamond_syn, name, home_dir)\n",
    "\n",
    "        for item in os.listdir(synbio):\n",
    "            if item.endswith('_profile'):\n",
    "                source = os.path.join(synbio, item)\n",
    "                destination = os.path.join(functional_folder, item)\n",
    "                shutil.move(source, destination)\n",
    "                \n",
    "        synbio_binary = '/home/anna/Documents/JGI_soil_genomes/functional_profiles/Synbio_functional_profile'\n",
    "        [distance_list_for_synbio, new_loc ]= pass_to_distance(synbio_binary, name, desired_location2, mg_to_analyze)\n",
    "        print(mg_to_analyze + 'vs. Synbio Analysis Complete')\n",
    "\n",
    "    \n",
    "    else: \n",
    "        print('\"They say that there\\'s no medicine that can cure a fool. I guess that\\'s true ...\" - Potion Shop Hag')\n",
    "        print('\\n Ocarina of time quotes aside, you entered something that isn\\'t available for analysis!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
