{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "#Initial Imports\n",
    "import gzip#; print('gzip version: ', gzip.__version__)\n",
    "import os#; print('os version: ', os.__version__)\n",
    "import re#; print('re version: ', re.__version__)\n",
    "import shutil#; print('shutil version: ', shutil.__version__)\n",
    "import requests#; print('requests version: ', requests.__version__)\n",
    "import wget#; print('wget version: ', wget.__version__)\n",
    "import subprocess#; print('subprocess version: ', subprocess.__version__)\n",
    "import time#; print('time version: ', time.__version__)\n",
    "from datetime import datetime\n",
    "from subprocess import PIPE, Popen\n",
    "from fake_useragent import UserAgent\n",
    "import os.path\n",
    "from os import path\n",
    "from Bio.ExPASy import Enzyme\n",
    "import numpy as np \n",
    "from sklearn.metrics import pairwise_distances\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "red = \"\\033[91m\"\n",
    "reset_color = \"\\033[0m\"\n",
    "\n",
    "\n",
    "from Encen_HPC import EC_extract, tsv_to_fasta, diamond_impl, ge\n",
    "#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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to Environmental Census (EnCen), a bioinformatics tool for characterizing synthetic biological risk to the environment \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('Welcome to Environmental Census (EnCen), a bioinformatics tool for characterizing synthetic biological risk to the environment \\n')\n",
    "EC_extract()\n",
    "tsv_to_fasta()\n",
    "mg_to_analyze = input('Please input the metagenome you would like to analyze:')\n",
    "#options are industrial wastewater, WWTP, and River "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"They say that there's no medicine that can cure a fool. I guess that\n",
      " true ...\" - Potion Shop Hag\n"
     ]
    }
   ],
   "source": [
    "if mg_to_analyze == 'soil':\n",
    "    metagenome_name = 'diamond_analysis_output' #-> folder\n",
    "    desired_location = '/home/anna/Documents/JGI_soil_genomes' \n",
    "    # Freshwater = nump/nump/nump/files\n",
    "    soil = '/home/anna/Documents/JGI_soil_genomes/Soil_Metagenome_Test'\n",
    "    abspath = os.path.abspath(soil)\n",
    "    name = 'soil_metagenome'\n",
    "\n",
    "#______________________________________________________________#\n",
    "    os.chdir(desired_location) #-> we are in the folder we want\n",
    "    if os.path.exists(metagenome_name):\n",
    "        shutil.rmtree(metagenome_name)\n",
    "        os.mkdir(metagenome_name)\n",
    "    else:#makes a new directory called metagenome_name\n",
    "        os.mkdir(metagenome_name)\n",
    "    desired_location = desired_location + \"/\" + metagenome_name \n",
    "# shutil.copy(abspath, desired_location) #moves  file to Test Cases folder\n",
    "#matches = metagenome + \"_matches\"\n",
    "    os.chdir(soil)\n",
    "    diamond = diamond_impl(soil, '') #-> Takes in the path and directory\n",
    "    final_folder = '/home/anna/Documents/JGI_soil_genomes/diamond_analysis_output'\n",
    "    desired_location = '/home/anna/Documents/JGI_soil_genomes/Soil_Metagenome_Test'\n",
    "    ff_name = 'functional_profiles'\n",
    "    functional_folder = '/home/anna/Documents/JGI_soil_genomes/functional_profiles'\n",
    "\n",
    "    # print(desired_location)\n",
    "    for item in os.listdir(desired_location):\n",
    "        if item.endswith(('_matches.tsv', '.dmnd')):\n",
    "            source = os.path.join(desired_location, item)\n",
    "            destination = os.path.join(final_folder, item)\n",
    "            shutil.move(source, destination)\n",
    "    output = genome_extractor(final_folder, name)\n",
    "\n",
    "    os.chdir('/home/anna/Documents/JGI_soil_genomes')\n",
    "    if os.path.exists(functional_folder):\n",
    "        shutil.rmtree(ff_name)\n",
    "        os.mkdir(ff_name)\n",
    "    else:#makes a new directory called metagenome_name\n",
    "        os.mkdir(ff_name)\n",
    "\n",
    "    for item in os.listdir(final_folder):\n",
    "        if item.endswith('_profile'):\n",
    "            source = os.path.join(final_folder, item)\n",
    "            destination = os.path.join(functional_folder, item)\n",
    "            shutil.move(source, destination)\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "elif mg_to_analyze == 'WWTP':\n",
    "    print('works')\n",
    "else: \n",
    "    print('\"They say that there\\'s no medicine that can cure a fool. I guess that\\n true ...\" - Potion Shop Hag')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
